src/agent/debugger/src/target.rs

```diff
@@ -479,17 +482,83 @@ impl Target {
     }
 
     pub fn prepare_to_resume(&mut self) -> Result<()> {
-        if let Some(context) = self.current_context.take() {
-            if self.context_is_modified {
-                context.set_thread_context(self.current_thread_handle)?;
+        // When resuming, we take extra care to avoid missing breakpoints. There are 2 race
+        // conditions considered here:
+        //
+        // * Threads A and B hit a breakpoint at literally the same time, so we have
+        //   multiple events already queued to handle those breakpoints.
+        //   Restoring the breakpoint for thread A should not interfere with restoring
+        //   the original code for thread B.
+        // * Thread A hit a breakpoint, thread B is 1 instruction away from hitting
+        //   that breakpoint. Thread B cannot miss the breakpoint when thread A is resumed.
+        //
+        // To avoid these possible races, when resuming, we only let a single thread go **if**
+        // we're single stepping any threads.
+        //
+        // First, if we last stepped because of hitting a breakpoint, we restore the breakpoint
+        // so that whichever thread is resumed, it can't miss the breakpoint.
+        if let Some(last_breakpoint_address) = self.last_breakpoint_address.take() {
+            trace!("Restoring breakpoint at 0x{:x}", last_breakpoint_address);
+            write_instruction_byte(self.process_handle, last_breakpoint_address, 0xcc)?;
+        }
+
+        if self.single_step.is_empty() {
+            // Resume all threads if we aren't waiting for any threads to single step.
+            trace!("Resuming all threads");
+
+            for thread_info in self.thread_info.values_mut() {
+                thread_info.resume_thread()?;
+            }
+        } else {
+            // Otherwise pick one thread to resume, giving preference to the current thread.
```

Comment:
Do we get any weird scheduling artifacts from this? With this strategy, does the prioritization of the current thread mean it gets to resume over and over, with its peers suspended, as long as it keeps hitting breakpoints?

Reply:
I'm not sure, but I removed it. It might have been causing weird issues that I couldn't root cause, but in further testing I haven't seen those same issues, so maybe it was bad idea.